### PR TITLE
Upgrade Certbot dependencies

### LINKED ISSUE #8030: Fix rebuild dependencies to include python version constraint (for enum)
We have code to do this, but it seems to no longer be working. This came up in the process of testing https://github.com/certbot/certbot/pull/8026. Until it's fixed, we have to remember to manually add the constraint whenever we run `rebuild_dependencies.py`.

### PR DESCRIPTION
**Pull request by adferrand**
This PR upgrades Certbot pinned dependencies through `letsencrypt-auto-source/rebuild_dependencies.py` while taking into account the problems detected in https://github.com/certbot/certbot/pull/8035:
* `cryptography` is pinned to `2.8` to continue to support OpenSSL 1.0.1 on non-x86 ancient Linux distributions (RHEL 6 + Debian 8)
* `parsedatetime` is pinned to `2.5` because of an incompatibility with Python 2.7 (see https://github.com/bear/parsedatetime/issues/246)
* `letsencrypt-auto-source/rebuild_dependencies.py` now takes into account the environment markers that are aded to `AUTHORITATIVE_CONSTRAINTS`: this is used for the `enum34` dependency, to not install it on Python 3.6+ and not break the distribution by swapping the built-in `enum` module during the setup of Certbot venv.

Fixes #8030

### CODE REVIEW

File: letsencrypt-auto-source/rebuild_dependencies.py [hunk 1]
```diff
@@ -42,13 +42,13 @@
     # Too touchy to move to a new version. And will be removed soon
     # in favor of pure python parser for Apache.
     'python-augeas': '0.5.0',
-    # Package enum34 needs to be explicitly limited to Python2.x, in order to avoid
-    # certbot-auto failures on Python 3.6+ which enum34 doesn't support. See #5456.
-    # TODO: hashin seems to overwrite environment markers in dependencies. This needs to be fixed.
-    'enum34': '1.1.6 ; python_version < \'3.4\'',
+    # Cryptography 2.9+ drops support for OpenSSL 1.0.1, but we still need to support it
```
Comment (bmw) [inline letsencrypt-auto-source/rebuild_dependencies.py:45]: nit:
```suggestion
    # Cryptography 2.9+ drops support for OpenSSL 1.0.1, but we still want to support it
```
I think as soon as we have a reason to upgrade it, we should just do so and move on.

**Comment by adferrand** [comment]
The tests at https://travis-ci.com/github/certbot/certbot/builds/169115564 were failing for recent systems that use Python3 (Fedora 30+, RHEL8) for the stdist farm tests. Indeed, in that situation were the sources are used and not the wheels, `enum34` was still breaking Python 3.7 (more precisely `distutils`) despite the fix done to not break Python 3.5+ on version 1.1.8 of `enum34`.

So I added back the pinning for `enum34`, on version `1.1.10`, and fixed `rebuild_dependencies.py` to inject the environment markers (here `python_version < '3.4'`) into the final `dependency-requirements.txt` file.

PR description is updated accordingly.

**Comment by adferrand** [comment]
Full test suite is passing now: https://travis-ci.com/github/certbot/certbot/builds/169179814

File: letsencrypt-auto-source/pieces/dependency-requirements.txt [hunk 2]
```diff
@@ -75,67 +70,66 @@ cryptography==2.8 \
...
-# certbot-auto failures on Python 3.6+ which enum34 doesn't support. See #5456.
-enum34==1.1.6 ; python_version < '3.4' \
-    --hash=sha256:2d81cbbe0e73112bdfe6ef8576f2238f2ba27dd0d55752a776c41d38b7da2850 \
-    --hash=sha256:644837f692e5f550741432dd3f223bbb9852018674981b1664e5dc339387588a \
-    --hash=sha256:6bd0f6ad48ec2aa117d3d141940d484deccda84d4fcd884f5c3d93c23ecd8c79 \
-    --hash=sha256:8ad8c4783bf61ded74527bffb48ed9b54166685e4230386a9ed9b1279e2df5b1
+distro==1.5.0 \
+    --hash=sha256:0e58756ae38fbd8fc3020d54badb8eae17c5b9dcbed388b17bb55b8a5928df92 \
+    --hash=sha256:df74eed763e18d10d0da624258524ae80486432cd17392d9c3d96f5e83cd2799
+enum34==1.1.10; python_version < '3.4' \
```
[the remark this answers is not in the document]
Comment (adferrand) [reply letsencrypt-auto-source/pieces/dependency-requirements.txt:76]: It seems that hashin is effectively parsing the requirement, and so reconstruct is in the requirement file with its own convention as coded in the tool:
```bash
$ touch test.txt
$ hashin enum34=="1.1.10 ; python < 3.4" -r test.txt
$ cat test.txt
enum34==1.1.10; python < 3.4 \
    --hash=sha256:a98a201d6de3f2ab3db284e70a33b0f896fbf35f8086594e8c9e74b909058d53 \
    --hash=sha256:c3858660960c984d6ab0ebad691265180da2b43f07e061c0f8dca9ef3cffd328 \
    --hash=sha256:cce6a7477ed816bd2542d03d53db9f0db935dd013b70f336a95c73979289f248
```

https://github.com/peterbe/hashin/blob/a9e2239bf46fcba12be9ecac272a806fc48bfe1e/hashin.py#L246-L254

To be coherent, I update the `AUTHORITATIVE_CONSTRAINTS` accordingly.

**Comment by FelixSchwarz** [comment]
May I suggest you also change the version requirements in `setup.py`? That way I can make sure all the Fedora/CentOS versions work fine.

For example if I push certbot 1.5 and sometime later `parsedatetime` is upgraded to 2.6 (#8042) we may not notice the problem. Requirements from `setup.py` will become RPM requirements so we can detect the incompatibility with semi-automated tooling in Fedora.

**Comment by bmw** [comment]
Yeah just to make sure we're all on the same page, I don't think we should do anything in `setup.py`. Like Felix said in their edit, `parsedatetime` is completely broken in Python 2. If you try to run the tests included in the package, they fail.

Because of that and other concerns I personally have with trying to work around things like this ourselves in `setup.py`, I don't think we should do anything here. If people have different opinions, I'm happy to chat, but I think we should probably do so in a separate issue/PR.

**Pull request merged**